hnn_core/gui/gui.py

```diff
@@ -398,7 +421,102 @@ def compose(self):
                                     self.widget_tstop, self._load_info,
                                     self.layout)
 
-        return hnn_gui
+        if not return_layout:
+            return
+        else:
+            return self.app_layout
+
+    def show(self):
+        display(self.app_layout)
+
+    def take_screenshot(self, width=None, height=None, filename=None,
+                        render=True):
+        """Embed the widget to HTML and render it through iframe.
+        Parameters
+        ----------
+        width : Optional[int]
+            The width of iframe window use to show the snapshot.
+        height : Optional[int]
+            The height of iframe window use to show the snapshot.
+        filename : Optional[int]
+            The filename of snapshot.
+        render: bool
+            Will return an IFrame object if False
+
+        Returns
+        -------
+        snapshot : An iframe snapshot object that can be rendered in notebooks.
+        """
+        self._screenshots_count += 1
+        if not filename:
+            filename = f"snapshot_{self._screenshots_count}.html"
+        embed_minimal_html(filename, views=[self.app_layout], title='')
```

Comment:
how come this works now? Was the trick to pass `self.app_layout` to `views` ?

Reply:
Inside `embed_minimal_html`, there is a function that extract data models from the widget list it received. Then that function turns each data model into a json object so that it can be pickled and later loaded. The actually stored objects are not html contents but a list of json objects.

Reply:
so is that how you made it work? I remember this was not working on Friday ...

Reply:
No on Friday the problem is more about the callback delay. Basically we bypassed that problem by sequentially execute all cells with JS